src/main.rs

```diff
@@ -140,19 +187,14 @@ struct MatchTree {
     branches: Vec<(ByteSet, MatchTree)>,
 }
 
-enum Cond {
-    Expr(Expr),
-    MatchTree(MatchTree),
-}
-
 /// Decoders with a fixed amount of lookahead
 enum Decoder {
     Fail,
     Empty,
     EndOfInput,
     Byte(ByteSet),
-    If(Cond, Box<Decoder>, Box<Decoder>),
-    Switch(MatchTree, Vec<Decoder>),
+    Alt(MatchTree, Box<Decoder>, Box<Decoder>),
+    Union(MatchTree, Vec<Decoder>),
```

Comment:
Maybe we can drop Alt altogether and rename Decoder::Union to Branch, to emphasise it's deterministic?